remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+				      struct strbuf *error_message)
+{
+	int ret = HTTP_OK;
+	const struct credential *p_creds = NULL;
+	long http_response_code;
+
+	lookup_main_creds();
+
+	if (rv_data.main_creds.username && *rv_data.main_creds.username)
+		p_creds = &rv_data.main_creds;
+
+	strbuf_setlen(error_message, 0);
+	trace2_region_enter("remote-vfs", "GET/object/main1", NULL);
+	ret = do_get__into_file(rv_data.buf_main_url.buf, url_component,
+				p_creds, tempfile,
+				&http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/main1", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+
+	if (ret != HTTP_REAUTH)
+		return ret;
+
+	refresh_main_creds();
+
+	/*
+	 * Since we are automatically retrying, truncate the error
+	 * message from the first attempt.  The caller should only
+	 * see the error message from the final attempt.
+	 */
+	strbuf_setlen(error_message, 0);
+
+	// TODO ftruncate tempfile ??
+
+	p_creds = &rv_data.main_creds;
+	trace2_region_enter("remote-vfs", "GET/object/main2", NULL);
+	ret = do_get__into_file(rv_data.buf_main_url.buf, url_component,
+				p_creds, tempfile,
+				&http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/main2", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+	return ret;
+}
+
+static int do_get__into_file__to_cache_server(const char *url_component,
+					      struct tempfile *tempfile,
+					      struct strbuf *error_message)
+{
+	long http_response_code;
+	int ret = HTTP_OK;
+
+	synthesize_cache_server_creds();
+
+	strbuf_setlen(error_message, 0);
+	trace2_region_enter("remote-vfs", "GET/object/cs1", NULL);
+	ret = do_get__into_file(rv_data.buf_cache_server_url.buf,
+				url_component, &rv_data.cache_creds,
+				tempfile, &http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/cs1", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+
+	if (http_response_code == 404)
+		return RV_ERROR__NOT_FOUND;
+
+	if (http_response_code != 400)
+		return RV_ERROR__ERROR;
+
+	/* Assume a 400 is a virtual-401 and retry */
+
+	refresh_cache_server_creds();
+
+	strbuf_setlen(error_message, 0);
+
+	// TODO ftruncate tempfile ??
+
+	trace2_region_enter("remote-vfs", "GET/object/cs2", NULL);
+	ret = do_get__into_file(rv_data.buf_cache_server_url.buf,
+				url_component, &rv_data.cache_creds,
+				tempfile, &http_response_code, error_message);
+	trace2_region_leave("remote-vfs", "GET/object/cs2", NULL);
+	if (ret == HTTP_OK)
+		return HTTP_OK;
+
+	if (http_response_code == 404)
+		return RV_ERROR__NOT_FOUND;
+
+	return ret;
+}
+
+static int do_get__into_file__with_fallback(const char *url_component,
+					    struct tempfile *tempfile,
+					    struct strbuf *error_message)
+{
+	int ret = HTTP_OK;
+
+	if (rv_data.buf_cache_server_url.len) {
+		ret = do_get__into_file__to_cache_server(
+			url_component, tempfile, error_message);
+		if (ret == HTTP_OK)
+			return ret;
+
+		if (!rv_opts.param__allow_fallback)
+			return ret;
```

Comment:
How about combining conditions into a single `if`, when the conditional action is the same anyway?

Reply:
combining a bunch of unrelated logic just to DRY a "return" statement IMHO obscures the meaning.  and again, makes it more difficult to set specific breakpoints.